### PR TITLE
Upgrade to bokeh 2 & robustify front-end tests

### PR DESCRIPTION
**Pull request by stefanv**
Closes https://github.com/skyportal/skyportal/issues/1250

### CODE REVIEW

**Comment by stefanv** [comment]
Please try the plots and see if anything is misbehaving.  I think I got most of them!

@kmshin1397 We may now also try and get the plots to scale dynamically in width; bokeh 0.12 didn't do that very well, but that was a long time ago.

**Review by dannygoldstein — COMMENTED** [review]
Dynamic width scaling would be great. I really like the compactified legends. One comment - looks like the slider bars are a little cut off on the spectrum and photometry plots. Also, in flux space, the slider bar on the photometry plot extends far to the right of the div, any way to truncate this to match the length of the slider in the mag plot? 
Screen Shot 2020-11-05 at 11 30 35 PM

**Comment by stefanv** [comment]
@dannygoldstein I can verify that behavior.  Given the complexity of this PR, I will probably take that one on in another PR.  Feel free to file an issue for it.

**Comment by stefanv** [comment]
I don't see the cut-off bars you mention (I do see the extended bar); do you mind posting a screenshot of that as well?

**Comment by stefanv** [comment]
@dannygoldstein I didn't realize the slider was so different than before; let me take a look.

**Comment by stefanv** [comment]
OK, @dannygoldstein, this should mostly address your concerns.  It is not perfect by any means, but I think it is better than what we used to have.

File: static/js/components/Candidate.jsx
Comment (kmshin1397) [inline static/js/components/Candidate.jsx:165]: There's a style rule on `classes.photometryContainer` (from `useSourceStyles()`) that says `padding-bottom: 0.5rem`. I found that changing that to `padding: 0.5rem` solves the problem of the slider bar in the photometry plot being cut off slightly on the left edge. 

Screenshot 2020-11-09 111154

**Review by kmshin1397 — COMMENTED** [review]
I haven't gotten through all the code yet, but some initial impressions from playing around with it.
- I think there is an indexing issue with the spectroscopy plots with the elements checkboxes. The colors that are plotted aren't matched up with the boxes that are selected.
Screenshot 2020-11-09 111418
- Probably a related issue, but it seems to me that the redshift/v_expansion sliders are only working for elements selected in the first three columns. 

My guess is that it's something hard-coded to 3 in the custom JSON for those. I'll try to look for the bug while I do the detailed code review.

File: static/js/components/Candidate.jsx
Comment (stefanv) [reply static/js/components/Candidate.jsx:165]: @kmshin1397 Thank you for helping with the review; feel free to push changes to this branch!

I think I know where the color mismatch came in---I re-oriented the checkboxes to save space.

**Comment by dannygoldstein** [comment]
@stefanv The dynamic scaling looks great - thanks for that. For the cut off sliders @kmshin1397 's comment should fix that. 

Any way to cut the size of the flux slider bar in half? Then will give green checkmark.

File: static/js/components/Candidate.jsx
Comment (kmshin1397) [reply static/js/components/Candidate.jsx:165]: @stefanv Yeah, it's probably easiest for me to fix that since I wrote that part originally. I hard-coded some stuff in the JS logic for those that isn't responsive to the number of columns you set. I'll put that on my list of things to do today.

Comment (stefanv) [reply static/js/components/Candidate.jsx:165]: > @kmshin1397 Thank you for helping with the review; feel free to push changes to this branch!
> 
> I think I know where the color mismatch came in---I re-oriented the checkboxes to save space.

Just to be clear: I meant I would not mind, not that you should!  I am happy to fix these myself later today.

Comment (kmshin1397) [reply static/js/components/Candidate.jsx:165]: @stefanv No problem, I think I see all the errors - will push a patch soon

**Comment by kmshin1397** [comment]
@stefanv I made a PR to your branch bc I don't have push access - once you merge that I'm good to approve this.

**Comment by stefanv** [comment]
@dannygoldstein Thanks to @kmshin1397's help, this should now be close to ready.

**Review by kmshin1397 — APPROVED** [review]

**Comment by dannygoldstein** [comment]
@stefanv Looks like several tests on travis are timing out waiting for the bokeh plots to load?

**Comment by stefanv** [comment]
They're probably waiting for divs that no longer render.

**Comment by dannygoldstein** [comment]
@stefanv 

If any of the tests look for text in the plot divs (e.g., text in the legend), then we will need to continue waiting for the plot divs to load in those tests (I think there may be one in the upload spectrum page, can't remember). This is because the bokeh plots are generally the last components to load on the page and can move other components around and interrupt selenium actions.

Otherwise, we could safely disable the plot divs.

**Comment by stefanv** [comment]
Right, so if we disable the plot generation we can remove those parts of the tests, since there would be no more moving around of components.

**Comment by dannygoldstein** [comment]
Except for the parts of the tests that look for text in the generated plots (e.g., legend text)

________________________________

**Comment by dannygoldstein** [comment]
@stefanv  any updates here? I think merging this will fix #955

**Review by acrellin — APPROVED** [review]

File: skyportal/plot.py
Comment (dannygoldstein) [inline skyportal/plot.py:745]: `plot.line -> plot.step`

Comment (acrellin) [reply skyportal/plot.py]: Done and pushed

**Review by dannygoldstein — COMMENTED** [review]
@acrellin really nice to see this almost there. Just a couple things - 

Missing colormap stuff from 

https://github.com/skyportal/skyportal/pull/1424/files#diff-25516b0881825b49b7c2b3d2721d148fc24c69b085a443458952693603cc7744L729-R729

Users requested retiring viridis due to the yellow colors (hard to see) and the sequential nature of the map, making it hard to tell one spectrum apart from the other (they blend into each other on plots with a lot of spectra). 

And permissions checking from 

https://github.com/skyportal/skyportal/pull/1396

File: skyportal/plot.py
Comment (dannygoldstein) [inline skyportal/plot.py:768]: this should be `spec_labels` as before, as the order of `spectra` is not the same as the order of `split`, so this results in the spectra being mislabeled

**Comment by acrellin** [comment]
@dannygoldstein thanks for catching those - pushed. Can you take another look?
Lesson learned: do not attempt to resolve gnarly merge conflicts for PRs you did not author :wink:

File: skyportal/plot.py
Comment (acrellin) [reply skyportal/plot.py]: Done

File: skyportal/plot.py
Comment (dannygoldstein) [inline skyportal/plot.py:780]:
```suggestion
        labels=spec_labels
```

Comment (acrellin) [reply skyportal/plot.py]: Ah I see thanks, I misunderstood. Pushed.

**Review by dannygoldstein — APPROVED** [review]
Looks like the changes to

**Comment by dannygoldstein** [comment]
is make load demo data failure real ?

**Review by kmshin1397 — APPROVED** [review]

**Comment by acrellin** [comment]
I had to restart twice, but load demo data passed eventually. It's taking a long time for the app to spin up apparently

**Comment by stefanv** [comment]
I have figured out the problem: if you try to connect to SkyPortal before it is ready, nginx marks it as unavailable.  In that case, you have to wait for the full timeout before nginx retries connecting you to the service.  I will see if I can tune our nginx configuration.

**Comment by stefanv** [comment]
This should help with the load_demo_data timeout: https://github.com/cesium-ml/baselayer/pull/180

**Comment by stefanv** [comment]
Failures unrelated; loaded locally and the plots seem to display correctly.

**Pull request merged**
